src/charm.py

```diff
@@ -0,0 +1,595 @@
+
+    _stored = StoredState()
+
+    service_token_path = Path("service_token")
+    repo_check_web_service_path = Path("/home/ubuntu/repo_policy_compliance_service")
+    repo_check_web_service_script = Path("src/repo_policy_compliance_service.py")
+    repo_check_systemd_service = Path("/etc/systemd/system/repo-policy-compliance.service")
+
+    def __init__(self, *args, **kargs) -> None:
+        """Construct the charm.
+
+        Args:
+            args: List of arguments to be passed to the `CharmBase` class.
+            kargs: List of keyword arguments to be passed to the `CharmBase`
+                class.
+        """
+        super().__init__(*args, **kargs)
+
+        self._event_timer = EventTimer(self.unit.name)
+
+        self._stored.set_default(
+            path=self.config["path"],  # for detecting changes
+            runner_bin_url=None,
+        )
+
+        self.proxies: ProxySetting = {}
+        if http_proxy := get_env_var("JUJU_CHARM_HTTP_PROXY"):
+            self.proxies["http"] = http_proxy
+        if https_proxy := get_env_var("JUJU_CHARM_HTTPS_PROXY"):
+            self.proxies["https"] = https_proxy
+        if no_proxy := get_env_var("JUJU_CHARM_NO_PROXY"):
+            self.proxies["no_proxy"] = no_proxy
+
+        self.service_token = None
+
+        self.on.define_event("reconcile_runners", ReconcileRunnersEvent)
+        self.on.define_event("update_runner_bin", UpdateRunnerBinEvent)
+
+        self.framework.observe(self.on.install, self._on_install)
+        self.framework.observe(self.on.upgrade_charm, self._on_upgrade_charm)
+        self.framework.observe(self.on.config_changed, self._on_config_changed)
+        self.framework.observe(self.on.reconcile_runners, self._on_reconcile_runners)
+        self.framework.observe(self.on.update_runner_bin, self._on_update_runner_bin)
+        self.framework.observe(self.on.stop, self._on_stop)
+
+        self.framework.observe(self.on.check_runners_action, self._on_check_runners_action)
+        self.framework.observe(self.on.reconcile_runners_action, self._on_reconcile_runners_action)
+        self.framework.observe(self.on.flush_runners_action, self._on_flush_runners_action)
+        self.framework.observe(self.on.update_runner_bin_action, self._on_update_runner_bin)
+
+    def _get_runner_manager(
+        self, token: Optional[str] = None, path: Optional[str] = None
+    ) -> Optional[RunnerManager]:
+        """Get a RunnerManager instance, or None if missing config.
+
+        Args:
+            token: GitHub personal access token to manager the runners with.
+            path: GitHub repository path in the format '<org>/<repo>', or the GitHub organization
+                name.
+
+        Returns:
+            A instance of RunnerManager if the token and path configuration can be found.
+        """
+        if token is None:
+            token = self.config["token"]
+        if path is None:
+            path = self.config["path"]
```

Comment:
Out of curiosity, is this not a charm state because of when this charm was written? :D
Also, I'm thinking if the `token`, `path` and `service_token` validation can be done at a higher level on the controller function so that `_get_runner_manager` can be purely about getting the `RunnerManager`.

Reply:
I agree. You can find an example of a similar situation on this [Synapse PR](https://github.com/canonical/synapse-operator/pull/6) where we have a configuration (server_name) that needs to be set before the deployment too.

Reply:
This was not a charm state when this charm was written. Also this function was already in this module.
This can be refactored.

Reply:
Major comments